### ISSUE TITLE
Invalid code generation using java.lang.Exception

### ISSUE THREAD

**Issue by thheller** [Day8/re-frame-10x#219]
https://github.com/Day8/re-frame-10x/blob/b989b08ddc75755df443b2d988ef4eaf50be912b/src/mranderson048/re_frame/v0v10v2/re_frame/trace.cljc#L57-L60

Reader conditionals cannot be used inside a macro since they are resolved during read time. On the JVM it will always use the `(catch Exception ...` `:clj` path. This goes by unnoticed in normal CLJS since it blindly accepts any symbol with a `.` in it without further checks. `shadow-cljs` validates this now to prevent such errors.

```
------ WARNING #1 --------------------------------------------------------------
 File: mranderson048/re_frame/v0v10v2/re_frame/events.cljc:61:11
--------------------------------------------------------------------------------
  58 |       (if *handling*
  59 |         (console :error "re-frame: while handling \"" *handling* "\", dispatch-sync was called for \"" event-v "\". You can't call dispatch-sync within an event handler.")
  60 |         (binding [*handling*  event-v]
  61 |           (trace/with-trace {:operation event-id
-----------------^--------------------------------------------------------------
 Use of undeclared Var mranderson048.re-frame.v0v10v2.re-frame.events/java
--------------------------------------------------------------------------------
  62 |                              :op-type   kind
  63 |                              :tags      {:event event-v}}
  64 |             (interceptor/execute event-v interceptors)))))))
  65 |
--------------------------------------------------------------------------------
```

The generated code runs fine until it actually runs into the `catch` block and tries to resolve `java.lang.Exception`.

```
try{try{return mranderson048.re_frame.v0v10v2.re_frame.interceptor.execute(event_v,interceptors);
}finally {if(mranderson048.re_frame.v0v10v2.re_frame.trace.is_trace_enabled_QMARK_()){
var end__32349__auto___35575 = mranderson048.re_frame.v0v10v2.re_frame.interop.now();
var duration__32350__auto___35576 = (end__32349__auto___35575 - new cljs.core.Keyword(null,"start","start",-355208981).cljs$core$IFn$_invoke$arity$1(mranderson048.re_frame.v0v10v2.re_frame.trace._STAR_current_trace_STAR_));
var seq__35541_35577 = cljs.core.seq(cljs.core.deref(mranderson048.re_frame.v0v10v2.re_frame.trace.trace_cbs));
var chunk__35542_35578 = null;
var count__35543_35579 = (0);
var i__35544_35580 = (0);
while(true){
if((i__35544_35580 < count__35543_35579)){
var vec__35547_35581 = chunk__35542_35578.cljs$core$IIndexed$_nth$arity$2(null,i__35544_35580);
var k__32351__auto___35582 = cljs.core.nth.cljs$core$IFn$_invoke$arity$3(vec__35547_35581,(0),null);
var cb__32352__auto___35583 = cljs.core.nth.cljs$core$IFn$_invoke$arity$3(vec__35547_35581,(1),null);
try{var G__35552_35584 = new cljs.core.PersistentVector(null, 1, 5, cljs.core.PersistentVector.EMPTY_NODE, [cljs.core.assoc.cljs$core$IFn$_invoke$arity$variadic(mranderson048.re_frame.v0v10v2.re_frame.trace._STAR_current_trace_STAR_,new cljs.core.Keyword(null,"duration","duration",1444101068),duration__32350__auto___35576,cljs.core.prim_seq.cljs$core$IFn$_invoke$arity$2([new cljs.core.Keyword(null,"end","end",-268185958),mranderson048.re_frame.v0v10v2.re_frame.interop.now()], 0))], null);
(cb__32352__auto___35583.cljs$core$IFn$_invoke$arity$1 ? cb__32352__auto___35583.cljs$core$IFn$_invoke$arity$1(G__35552_35584) : cb__32352__auto___35583.call(null,G__35552_35584));
}catch (e35550){if((e35550 instanceof java.lang.Exception)){
var e__32353__auto___35585 = e35550;
mranderson048.re_frame.v0v10v2.re_frame.loggers.console.cljs$core$IFn$_invoke$arity$variadic(new cljs.core.Keyword(null,"error","error",-978969032),cljs.core.prim_seq.cljs$core$IFn$_invoke$arity$2(["Error thrown from trace cb",k__32351__auto___35582,"while storing",mranderson048.re_frame.v0v10v2.re_frame.trace._STAR_current_trace_STAR_,e__32353__auto___35585], 0));
} else {
throw e35550;

}
}
```

**Comment by danielcompton** [comment]
Ouch, sorry about that. Do you have any thoughts on how we could avoid this problem? Maybe making more use of macrovich? It's a shame that there aren't reader conditionals for target platform as well as for the executing platform.

**Comment by thheller** [comment]
I don't think this could be solved nicely via reader conditionals, at least not without extending that mechanism in some way.

The most commonly used trick I've seen to date involves checking some macro `&env` data that is only present during CLJS compilation.

```
(defmacro finish-trace [trace]
  `(when (is-trace-enabled?)
     (let [end#      (interop/now)
           duration# (- end# (:start ~trace))]
       (doseq [[k# cb#] @trace-cbs]
         (try (cb# [(assoc ~trace
                      :duration duration#
                      :end (interop/now))])
              (catch ~(if (:ns &env) :default 'Exception) e#
                (console :error "Error thrown from trace cb" k# "while storing" ~trace e#)))))))
```

**Comment by thheller** [comment]
This was fixed in https://github.com/Day8/re-frame/commit/3de2ff8ff3403af2fcf29f287462a36f58c7bbb0#diff-87c1083a777fb7472aa850b82f8c2def so just bumping the `re-frame` version via mranderson should be enough.

**Comment by danielcompton** [comment]
This is fixed in the recently released 0.3.4.
